src/H5Eint.c

```diff
@@ -98,6 +120,738 @@ char H5E_mpi_error_str[MPI_MAX_ERROR_STRING];
+ */
+ssize_t
+H5E__get_class_name(const H5E_cls_t *cls, char *name, size_t size)
+{
+    ssize_t len = -1; /* Length of error class's name */
+
+    FUNC_ENTER_PACKAGE_NOERR
+
+    /* Check arguments */
+    assert(cls);
+
+    /* Get the class's name */
+    len = (ssize_t)strlen(cls->cls_name);
+
+    /* Set the user's buffer, if provided */
+    if (name) {
+        strncpy(name, cls->cls_name, size);
+        if ((size_t)len >= size)
+            name[size - 1] = '\0';
+    } /* end if */
+
+    /* Return the full length */
+    FUNC_LEAVE_NOAPI(len)
+} /* end H5E__get_class_name() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5E__close_msg_cb
+ *
+ * Purpose:     H5I_iterate callback function to close error messages in the
+ *              error class.
+ *
+ * Return:      Success:    H5_ITER_CONT (0)
+ *              Failure:    H5_ITER_ERROR (-1)
+ *
+ *-------------------------------------------------------------------------
+ */
+static int
+H5E__close_msg_cb(void *obj_ptr, hid_t obj_id, void *udata)
+{
+    H5E_msg_t *err_msg   = (H5E_msg_t *)obj_ptr;
+    H5E_cls_t *cls       = (H5E_cls_t *)udata;
+    int        ret_value = H5_ITER_CONT; /* Return value */
+
+    FUNC_ENTER_PACKAGE
+
+    /* Check arguments */
+    assert(err_msg);
+
+    /* Close the message if it is in the class being closed */
+    if (err_msg->cls == cls) {
+        if (H5E__close_msg(err_msg, NULL) < 0)
+            HGOTO_ERROR(H5E_ERROR, H5E_CANTCLOSEOBJ, H5_ITER_ERROR, "unable to close error message");
+        if (NULL == H5I_remove(obj_id))
+            HGOTO_ERROR(H5E_ERROR, H5E_CANTREMOVE, H5_ITER_ERROR, "unable to remove error message");
+    } /* end if */
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5E__close_msg_cb() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5E__free_msg
+ *
+ * Purpose:     Private function to free an application error message.
+ *
+ * Return:      none
+ *
+ *-------------------------------------------------------------------------
+ */
+static void
+H5E__free_msg(H5E_msg_t *msg)
+{
+    FUNC_ENTER_PACKAGE_NOERR
+
+    /* Check arguments */
+    assert(msg);
+    assert(msg->app_msg);
+
+    /* Free resources */
+    msg->msg = H5MM_xfree_const(msg->msg);
+    msg      = H5FL_FREE(H5E_msg_t, msg);
+
+    FUNC_LEAVE_NOAPI_VOID
+} /* end H5E__free_msg() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5E__close_msg
+ *
+ * Purpose:     Private function to close an error message.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5E__close_msg(H5E_msg_t *err, void H5_ATTR_UNUSED **request)
+{
+    FUNC_ENTER_PACKAGE_NOERR
+
+    /* Check arguments */
+    assert(err);
+
+    /* Free resources, if application registered this message */
+    if (err->app_msg)
```

Comment:
Only application error messages need to be freed, the library error messages are statically declared variables.